core/src/connection/manager.rs

```diff
@@ -0,0 +1,479 @@
+        /// The connection ID.
+        ///
+        /// As a result of the error, the pending connection has been removed
+        /// from the `Manager` and is being closed. Hence this ID will
+        /// no longer resolve to a valid entry in the manager.
+        id: ConnectionId,
+        /// What happened.
+        error: ConnectionError<HE, TE>,
+        /// The handler that was supposed to handle the failed connection.
+        handler: H
+    },
+
+    /// An established connection has encountered an error.
+    ConnectionError {
+        /// The connection ID.
+        ///
+        /// As a result of the error, the connection has been removed
+        /// from the `Manager` and is being closed. Hence this ID will
+        /// no longer resolve to a valid entry in the manager.
+        id: ConnectionId,
+        /// Information about the connection that encountered the error.
+        connected: Connected<C>,
+        /// The error that occurred.
+        error: ConnectionError<HE, TE>,
+    },
+
+    /// A connection has been established.
+    ConnectionEstablished {
+        /// The entry associated with the new connection.
+        entry: EstablishedEntry<'a, I, C>,
+    },
+
+    /// A connection handler has produced an event.
+    ConnectionEvent {
+        /// The entry associated with the connection that produced the event.
+        entry: EstablishedEntry<'a, I, C>,
+        /// The produced event.
+        event: O
+    }
+}
+
+impl<I, O, H, TE, HE, C> Manager<I, O, H, TE, HE, C> {
+    /// Creates a new connection manager.
+    pub fn new(executor: Option<Box<dyn Executor + Send>>) -> Self {
+        let (tx, rx) = mpsc::channel(1);
+        Self {
+            tasks: FnvHashMap::default(),
+            next_task_id: TaskId(0),
+            executor,
+            local_spawns: FuturesUnordered::new(),
+            events_tx: tx,
+            events_rx: rx
+        }
+    }
+
+    /// Adds to the manager a future that tries to reach a node.
+    ///
+    /// This method spawns a task dedicated to resolving this future and
+    /// processing the node's events.
+    pub fn add_pending<F, M>(&mut self, future: F, handler: H) -> ConnectionId
+    where
+        I: Send + 'static,
+        O: Send + 'static,
+        TE: error::Error + Send + 'static,
+        HE: error::Error + Send + 'static,
+        C: Send + 'static,
+        M: StreamMuxer + Send + Sync + 'static,
+        M::OutboundSubstream: Send + 'static,
+        F: Future<Output = ConnectResult<C, M, HE, TE>> + Send + 'static,
+        H: IntoConnectionHandler<C> + Send + 'static,
+        H::Handler: ConnectionHandler<
+            Substream = Substream<M>,
+            InEvent = I,
+            OutEvent = O,
+            Error = HE
+        > + Send + 'static,
+        <H::Handler as ConnectionHandler>::OutboundOpenInfo: Send + 'static,
+    {
+        let task_id = self.next_task_id;
+        self.next_task_id.0 += 1;
+
+        let (tx, rx) = mpsc::channel(4);
```

Comment:
Out of curiosity, what is the number 4 based on? Why not just `channel(0)`, given that each sender is guaranteed a slot?

Reply:
Good question, I don't know! This is taken from pre-existing code introduced in https://github.com/libp2p/rust-libp2p/pull/1191 so maybe @twittner can shed more light on it. My naive guess would be that it is simply to not require complete "fairness" from the executor in polling the task that runs code using functions of the manager vs polling the background tasks of the connection(s) to consume the commands sent to the task(s), i.e. such a buffer limit of N could allow one task to take N steps before an executor is forced to switch to another task to make progress, in a sense controlling how far ahead a producer task can get before execution must continue on the consumer side. That is just a guess though.

Reply:
> given that each sender is guaranteed a slot?

We never clone this sender.

And yes, it is so that we can push multiple events without having to wait for the background task to make progress.

I don't think there's any drawback in increasing this value if we wanted to, except for the increase in memory usage. `4` is kind of arbitrary, but seems like a good value to me: unlikely to be reached in a normal situation, but small enough to not be a burden in memory.